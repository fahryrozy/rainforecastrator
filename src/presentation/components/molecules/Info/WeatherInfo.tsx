import {View, Text, Image} from 'react-native';
import React from 'react';
import moment from 'moment';
import 'moment/locale/id';
import {styles} from './CurrentWeather.styles';
import Clock from 'react-live-clock';

const WeatherInfo = ({data}) => {
  moment.locale('id');
  return (
    <View style={styles.mainContainer}>
      {data && (
        <View style={styles.weatherInfoCard}>
          <View style={styles.locCard}>
            <Text ellipsizeMode="tail" numberOfLines={1} style={styles.locText}>
              {data?.location?.name}
            </Text>
            <View style={styles.locTimeCard}>
              <Text style={styles.locTime}>Local time : </Text>
              {data?.location?.tz_id && (
                <Clock
                  style={styles.locTime}
                  format={'HH:mm:ss'}
                  ticking={true}
                  element={Text}
                  timezone={data.location.tz_id}
                />
              )}
            </View>
          </View>

          <View style={styles.tempCard}>
            <View style={styles.iconContainer}>
              <Image
                style={styles.icon}
                source={{
                  uri: `https:${data?.current?.condition?.icon}`,
                }}
              />
              <Text style={styles.tempText}>{`${data?.current?.temp_c}Â°`}</Text>
            </View>

            <Text style={styles.tempInfoText}>
              {data?.current?.condition?.text}
            </Text>

            <Text style={styles.lastUpdDate}>{`Last updated on ${moment(
              data?.current?.last_updated,
              'YYYY-MM-DD H:m',
            ).format('LT')}`}</Text>
          </View>
        </View>
      )}
    </View>
  );
};

export default WeatherInfo;
