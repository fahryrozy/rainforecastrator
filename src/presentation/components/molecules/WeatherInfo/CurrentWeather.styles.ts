import {Colors, scale} from '@core/style';
import {StyleSheet} from 'react-native';

export const styles = StyleSheet.create({
  weatherInfoCard: {
    flex: 1,
    justifyContent: 'flex-start',
    alignItems: 'flex-start',
    overflow: 'hidden',
    rowGap: scale(2),
  },
  iconContainer: {
    height: scale(75),
    justifyContent: 'center',
    alignItems: 'flex-start',
    flexDirection: 'row',
    width: '100%',
    overflow: 'hidden',
    gap: scale(10),
  },
  locCard: {
    flex: 0.35,
    rowGap: scale(2),
    justifyContent: 'center',
    alignItems: 'center',
    width: '100%',
    overflow: 'hidden',
    paddingVertical: scale(5),
  },
  tempCard: {
    flex: 0.65,
    paddingVertical: scale(2),
    justifyContent: 'flex-end',
    alignItems: 'center',
    width: '100%',
    overflow: 'hidden',
  },
  locTimeCard: {
    flexDirection: 'row',
    width: '100%',
    alignItems: 'center',
    justifyContent: 'center',
    overflow: 'hidden',
  },
  locTime: {
    textAlign: 'center',
    fontSize: scale(10),
    color: Colors.pureWhite,
    fontFamily: 'Poppins',
    fontWeight: '400',
  },
});
