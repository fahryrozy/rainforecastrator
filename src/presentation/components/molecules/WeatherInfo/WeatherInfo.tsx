import {View, Text, Image, ActivityIndicator} from 'react-native';
import React from 'react';
import moment from 'moment';
import 'moment/locale/id';
import {styles} from './CurrentWeather.styles';
import Clock from 'react-live-clock';
import {Current, Location} from '@domain/entities';
import * as Sentry from '@sentry/react';

type Props = {
  data: {
    location: Location;
    current: Current;
  };
  isLoading: boolean;
};

const WeatherInfo: React.FC<Props> = ({data, isLoading}) => {
  moment.locale('id');
  if (isLoading) {
    return (
      <ActivityIndicator
        size="large"
        color="#fff"
        style={styles.mainContainer}
      />
    );
  }
  return (
    <View style={styles.mainContainer}>
      {data && (
        <View style={styles.weatherInfoCard}>
          <View style={styles.locCard}>
            <Text ellipsizeMode="tail" numberOfLines={1} style={styles.locText}>
              {data?.location?.name}
            </Text>
            <View style={styles.locTimeCard}>
              <Text style={styles.locTime}>Local time : </Text>
              {data?.location?.tz_id ? (
                <Clock
                  style={styles.locTime}
                  format={'HH:mm:ss'}
                  ticking={true}
                  element={Text}
                  timezone={data.location.tz_id}
                />
              ) : (
                <Text>N/A</Text>
              )}
            </View>
          </View>

          <View style={styles.tempCard}>
            <View style={styles.iconContainer}>
              <Image
                style={styles.icon}
                source={{
                  uri: `https:${data?.current?.condition?.icon}`,
                }}
              />
              <Text style={styles.tempText}>{`${data?.current?.temp_c}Â°`}</Text>
            </View>

            <Text style={styles.tempInfoText}>
              {data?.current?.condition?.text}
            </Text>

            <Text style={styles.lastUpdDate}>{`Last updated on ${moment(
              data?.current?.last_updated,
              'YYYY-MM-DD H:m',
            ).format('LT')}`}</Text>
          </View>
        </View>
      )}
    </View>
  );
};

export default Sentry.withProfiler(WeatherInfo);
