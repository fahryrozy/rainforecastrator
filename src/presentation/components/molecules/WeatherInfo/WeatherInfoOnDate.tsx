import {View, Text, Image, ActivityIndicator} from 'react-native';
import React from 'react';
import moment from 'moment';
import 'moment/locale/id';
import {styles} from './CurrentWeather.styles';
import Clock from 'react-live-clock';
import {Forecast, Hour, Location} from '@domain/entities';
import {Colors, Typhography} from '@core/style';
import {Wrapper} from '@core/style/Wrapper';
import IconStyle from '@core/style/Icon';

type Props = {
  data:
    | {
        location: Location;
        forecast: Forecast;
      }
    | undefined;
  condition: Hour | undefined;
  isLoading: boolean;
};

const WeatherInfoOnDate: React.FC<Props> = ({data, condition, isLoading}) => {
  moment.locale('en');

  if (isLoading) {
    return <ActivityIndicator size="large" color={Colors.pureWhite} />;
  }
  return (
    <View style={[styles.weatherContainer, Wrapper.ContainerColumn()]}>
      {data && condition && (
        <View style={styles.weatherInfoCard}>
          <View style={styles.locCard}>
            <Text
              ellipsizeMode="tail"
              numberOfLines={1}
              style={Typhography.headerBigDefault()}>
              {data?.location?.name}
            </Text>
            <View style={styles.locTimeCard}>
              <Text style={Typhography.bodyDefault()}>Local time : </Text>
              {data?.location?.tz_id ? (
                <Clock
                  style={styles.locTime}
                  format={'HH:mm:ss'}
                  ticking={true}
                  element={Text}
                  timezone={data.location.tz_id}
                />
              ) : (
                <Text style={Typhography.bodyDefault()}>N/A</Text>
              )}
            </View>
          </View>

          <View style={styles.tempCard}>
            <View style={styles.iconContainer}>
              <Image
                style={IconStyle.iconBig}
                source={{
                  uri: `https:${condition.condition?.icon}`,
                }}
              />
              <Text
                style={Typhography.headerExtraBigDefault()}>{`${condition.temp_c}Â°`}</Text>
            </View>

            <Text style={Typhography.headerHalfBigDefault()}>
              {data?.forecast.forecastday[0]?.hour[0]?.condition?.text}
            </Text>

            <Text style={Typhography.bodyBold()}>{`${moment(
              condition.time,
              'YYYY-MM-DD H:m',
            ).format('llll')}`}</Text>
          </View>
        </View>
      )}
    </View>
  );
};

export default WeatherInfoOnDate;
