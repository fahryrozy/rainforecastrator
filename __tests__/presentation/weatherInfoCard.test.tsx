/**
 * @format
 */

import 'react-native';
import React from 'react';
import {it} from '@jest/globals';
import renderer from 'react-test-renderer';
import {render, screen} from '@testing-library/react-native';
import '@testing-library/jest-dom';
import '@testing-library/react-native/extend-expect';
import WeatherInfo from '@presentation/components/molecules/WeatherInfo/WeatherInfo';
import {mockData_weatherInfo} from '../data/mock';

describe('Weather Info Card in Homescreen', () => {
  it('renders correctly with blank data', () => {
    renderer.create(<WeatherInfo data={undefined} isLoading={false} />);
  });

  it('renders correctly with mock data', () => {
    renderer.create(
      <WeatherInfo data={mockData_weatherInfo} isLoading={false} />,
    );
  });

  it('match snapshot', () => {
    const snap = renderer
      .create(<WeatherInfo data={mockData_weatherInfo} isLoading={false} />)
      .toJSON();
    expect(snap).toMatchSnapshot();
  });

  it('has location information', () => {
    render(<WeatherInfo data={mockData_weatherInfo} isLoading={false} />);
    const titleText = screen.getByTestId('wi-locInfo');
    expect(titleText).toBeOnTheScreen();
  });

  it('has temp information', () => {
    render(<WeatherInfo data={mockData_weatherInfo} isLoading={false} />);
    const titleText = screen.getByTestId('wi-tempInfo');
    expect(titleText).toBeOnTheScreen();
  });

  it('has weather image', () => {
    render(<WeatherInfo data={mockData_weatherInfo} isLoading={false} />);
    const titleText = screen.getByTestId('wi-weatherImage');
    expect(titleText).toBeOnTheScreen();
  });
});
