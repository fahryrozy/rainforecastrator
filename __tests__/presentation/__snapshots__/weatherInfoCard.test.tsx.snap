// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Weather Info Card in Homescreen match snapshot 1`] = `
<View
  style={
    {
      "height": 400,
      "marginVertical": 5,
    }
  }
>
  <View
    style={
      {
        "alignItems": "flex-start",
        "flex": 1,
        "justifyContent": "flex-start",
        "overflow": "hidden",
        "rowGap": 4,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flex": 0.35,
          "justifyContent": "center",
          "overflow": "hidden",
          "paddingVertical": 10,
          "rowGap": 4,
          "width": "100%",
        }
      }
    >
      <Text
        ellipsizeMode="tail"
        numberOfLines={1}
        style={
          {
            "color": "rgba(255, 255, 255, 1)",
            "fontFamily": "Poppins",
            "fontSize": 68,
            "fontWeight": "400",
            "textAlign": "center",
            "width": "100%",
          }
        }
        testID="wi-locInfo"
      >
        Jakarta
      </Text>
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "center",
            "overflow": "hidden",
            "width": "100%",
          }
        }
      >
        <Text
          style={
            {
              "color": "rgba(255, 255, 255, 1)",
              "fontFamily": "Poppins",
              "fontSize": 20,
              "fontWeight": "400",
              "textAlign": "center",
            }
          }
        >
          Local time : 
        </Text>
        <Text
          dateTime={"2023-12-04T09:14:30.546Z"}
          style={
            {
              "color": "rgba(255, 255, 255, 1)",
              "fontFamily": "Poppins",
              "fontSize": 20,
              "fontWeight": "400",
              "textAlign": "center",
            }
          }
        >
          16:14:30
        </Text>
      </View>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "color": "rgba(255, 255, 255, 1)",
          "flex": 0.65,
          "justifyContent": "flex-end",
          "overflow": "hidden",
          "paddingVertical": 4,
          "width": "100%",
        }
      }
    >
      <View
        style={
          {
            "alignItems": "flex-start",
            "flexDirection": "row",
            "gap": 20,
            "height": 150,
            "justifyContent": "center",
            "overflow": "hidden",
            "width": "100%",
          }
        }
      >
        <Image
          source={
            {
              "uri": "https://cdn.weatherapi.com/weather/64x64/day/386.png",
            }
          }
          style={
            {
              "height": 150,
              "width": 150,
            }
          }
          testID="wi-weatherImage"
        />
        <Text
          style={
            {
              "color": "rgba(255, 255, 255, 1)",
              "fontFamily": "Poppins",
              "fontSize": 130,
              "textAlign": "center",
            }
          }
        >
          28Â°
        </Text>
      </View>
      <Text
        style={
          {
            "alignItems": "flex-start",
            "color": "rgba(255, 255, 255, 1)",
            "fontFamily": "Poppins",
            "fontSize": 40,
            "textAlign": "center",
          }
        }
        testID="wi-tempInfo"
      >
        Patchy light rain with thunder
      </Text>
      <Text
        style={
          {
            "color": "rgba(255, 255, 255, 1)",
            "flex": 1,
            "fontFamily": "Poppins",
            "fontSize": 20,
            "fontWeight": "400",
            "textAlign": "center",
          }
        }
      >
        Last updated on 13.45
      </Text>
    </View>
  </View>
</View>
`;
